patches/server/1054-Fix-DamageSource-API.patch

```diff
@@ -0,0 +1,166 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Sat, 9 Mar 2024 14:13:04 -0800
+Subject: [PATCH] Fix DamageSource API
+
+Uses the correct entity in the EntityDamageByEntity event
+Returns the correct entity for API's DamageSource#getCausingEntity
+
+diff --git a/src/main/java/net/minecraft/world/damagesource/DamageSource.java b/src/main/java/net/minecraft/world/damagesource/DamageSource.java
+index 1561b85a45f52a8162f43553f8485bfe084b8f1f..3f3bff037fc4ea9cb0aa9069d386e6f6922ee769 100644
+--- a/src/main/java/net/minecraft/world/damagesource/DamageSource.java
++++ b/src/main/java/net/minecraft/world/damagesource/DamageSource.java
+@@ -27,7 +27,8 @@ public class DamageSource {
+     private boolean withSweep = false;
+     private boolean melting = false;
+     private boolean poison = false;
+-    private Entity customCausingEntity = null; // This field is a helper for when causing entity damage is not set by vanilla
++    @Nullable
++    private Entity customEventDamager = null; // This field is a helper for when causing entity damage is not set by vanilla // Paper - fix DamageSource API
+ 
+     public DamageSource sweep() {
+         this.withSweep = true;
+@@ -56,13 +57,15 @@ public class DamageSource {
+         return this.poison;
+     }
+ 
+-    public Entity getCausingEntity() {
+-        return (this.customCausingEntity != null) ? this.customCausingEntity : this.causingEntity;
++    // Paper start - fix DamageSource API
++    public @Nullable Entity getCustomEventDamager() {
++        return (this.customEventDamager != null) ? this.customEventDamager : this.directEntity;
```

Comment:
this now ignore the causingEntity?

Reply:
Yeah, because the Event#getDamager should always return the direct entity if a custom one hasn't been set.

Reply:
That is the whole point.
You would not want the EntityDamageByEntityEvent to have the player lighting the tnt as the event damager.
You'd want the tnt itself as damager.

Reply:
yeah the thing dont remember correctly if the directEntity can be null.